### DISCUSSION TITLE
Handling Kernel Launches with Long Boot Times

### DISCUSSION THREAD

**Discussion by nicholaswold** [nteract/archived-desktop-app#5324] [Q&A]
Hi, was hoping to get some guidance. I’m developing an application that deals with kernels that have longer boot times, and when I lazy-load a kernel the `launchWebSocketKernelEpic` successfully creates a new session on my Jupyter Server, which returns an `execution_state: "starting"`. This in turn triggers this check in the `sendExecuteRequestEpic`, which sees that the execution state is "starting" and fails.

What I’d like to do is have the `launchWebSocketKernelEpic` poll the sessions API until the kernel is ready, or to wait until a `kernel_info_reply` message is sent on the websocket and then pull the latest execution state. I don’t see a way to do either right now. Is there another mechanism to accomplish what I’m trying to do, or is this additional functionality I can contribute to `launchWebSocketKernelEpic`?

**Comment by captainsafia** [comment]
@nicholaswold Sorry for the delay!

First, the situation that you are running into is interesting. In particular, I would expect the kernel to eventually transition from starting to connected state once the `kernel_info_reply` messages have been processed. It seems from your explanation that that isn't happening so we might have a bug here.

To address this, we probably need to add a `setKernelStatus` action [here](https://github.com/nteract/nteract/blob/main/packages/epics/src/kernel-lifecycle.ts#L143).

This is my preliminary analysis of this. Although I'm 🤔 about what caused the regression here since this behavior usually works.

**Reply by nicholaswold** [reply]
100% agree with the expectation. Looking at my debugging tools I can see the kernel info get updated when the `kernel_info_reply` returns, but it doesn't seem that the `launchWebSocketKernelEpic` waits for that to happen before signaling that the kernel had successfully launched. [This is the line of code that makes me think that it fires off an info request without waiting for a reply](https://github.com/nteract/nteract/blob/main/packages/epics/src/websocket-kernel.ts#L89). I'm still muddling through React-land though so I might just not understand what that line of code does.

For reference, here's what I got from Redux Devtools that got me to this conclusion:

![image](https://user-images.githubusercontent.com/1581984/96170709-ca9ddc00-0ed8-11eb-915f-9c5734b4c956.png)

I expected that it would wait for the reply with some sort of timeout, similar to what the restart epics do, and then move forward once it's sure the kernel came up. Does that make sense?

**Reply by nicholaswold** [reply]
Oh, I just double-checked and you're right [this error](https://github.com/nteract/nteract/blob/49983455451c1038a59408537e8698ebca9f074a/packages/epics/src/execute/execute.ts#L400) fires even after `CORE/SET_KERNEL_INFO` resolves. That makes me think that the kernel status never gets updated in this case 🤔 

In my case, I click the "run cell" button in my app and it keeps throwing that error. If I refresh, it reestablishes the connection, and since the kernel process is already alive it has no issues.

**Comment by nicholaswold** [comment]
@captainsafia I've got some time today to implement your suggested fix for `acquireKernelInfo`, I'll tag you on the PR.

Let me know if you have any ideas about the other issue I'm having about Nteract firing `launchKernelSuccessful` before the kernel is actually ready. I'm a little hesitant to start hacking away at that to solve for my use-case because I don't know if I'd be changing expected behavior there.

**Comment by captainsafia** [comment]
@nicholaswold Thanks for digging into this and apologies for the delay.

> This is the line of code that makes me think that it fires off an info request without waiting for a reply.

I think this is fine. In the worst case scenario, we fire off the message but don't get a response back from the kernel (because something was wrong with it) and our "kernel not connected" logic kicks in.

In the best-case scenario, we process the `kernel_info_reply` in the epic above but don't actually bother to update the kernel status. We should use the fact that we got a `kernel_info_reply` as an indicator that the kernel has been connected successfully.

> @captainsafia I've got some time today to implement your suggested fix for acquireKernelInfo, I'll tag you on the PR.

Yep! This sounds sensible. Thanks!
